trestle/core/control_io.py

```diff
@@ -807,71 +737,236 @@ def _read_sections(ii: int, lines: List[str], control_id: str,
-                if ii < 0:
-                    break
+            yaml_header, control = md_api.processor.process_markdown(control_file)
+
+            imp_string = 'Implementation'
+            headers = control.get_all_headers_for_key(imp_string, False)
+            if not headers:
+                # if statement has no parts there is only one response for entire control
+                headers = control.get_all_headers_for_key(const.SSP_MD_IMPLEMENTATION_QUESTION, False)
+                # should be only one header
+                for header in headers:
+                    node = control.get_node_for_key(header)
+                    ControlIOReader._add_node_to_dict(comp_name, 'Statement', comp_dict, node)
+            else:
+                for header in headers:
+                    tokens = header.split(' ', 2)
+                    if tokens[0] == '##' and tokens[1] == imp_string:
+                        label = tokens[2].strip()
+                        node = control.get_node_for_key(header)
+                        ControlIOReader._add_node_to_dict(comp_name, label, comp_dict, node)
+
         except TrestleError as e:
             logger.error(f'Error occurred reading {control_file}')
             raise e
-        return responses, header
+        return comp_dict, yaml_header
 
     @staticmethod
-    def read_implementations(control_file: pathlib.Path,
-                             component: ossp.SystemComponent) -> List[ossp.ImplementedRequirement]:
-        """Get implementation requirements associated with given control and link to the one component we created."""
+    def _insert_header_content(imp_req: ossp.ImplementedRequirement, header: Dict[str, Any]) -> None:
+        """Insert yaml header content into the imp_req and its by_comps."""
+        dict_ = header.get(const.SSP_FEDRAMP_TAG, {})
+        control_orig = dict_.get(const.CONTROL_ORIGINATION, [])
+        imp_status = dict_.get(const.IMPLEMENTATION_STATUS, [])
+        roles = dict_.get(const.RESPONSIBLE_ROLES, [])
+        props = []
+        responsible_roles = []
+        for co in control_orig:
+            if isinstance(co, str):
+                props.append(common.Property(ns=const.NAMESPACE_FEDRAMP, name=const.CONTROL_ORIGINATION, value=co))
+            elif isinstance(co, dict):
+                if const.INHERITED in co:
+                    uuid = co[const.INHERITED]
+                    props.append(common.Property(name=const.LEV_AUTH_UUID, value=uuid))
+                    props.append(
+                        common.Property(
+                            ns=const.NAMESPACE_FEDRAMP, name=const.CONTROL_ORIGINATION, value=const.INHERITED
+                        )
+                    )
+                else:
+                    raise TrestleError(f'Unexpected dict in control yaml header: {co}')
+            else:
+                raise TrestleError(f'Unexpected structure in control yaml header: {co}')
```

Comment:
These error messages are going to be indecipherable to people. I think you need some more verbose logging / error handling here.

Reply:
I will pass in the control id at least and try to elaborate - but if the content is unexpected then I don't have much other insight about it except that it is unexpected.  But I will make the msg friendlier.